### ISSUE TITLE
[Feature] Fail suite on retried fail

### ISSUE THREAD

**Issue by dimkin-eu** [microsoft/playwright#14215]
As for now, there are 2 options available for retries and fast fails
`--retries=1 --max-failures=1`

If the test is flaky even 1st fail will stop the suite. setting `--max-failures=2`, will fail on 2 flaky tests failures, which retry would pass.

Could be nice if `--max-failures=1` would fail on hard fail, i.e. flaky test failed also in retry

**Comment by pavelfeldman** [comment]
I started looking into implementing it and stumbled upon something: I was thinking of moving retries to the very end - this strategy worked well for us in Chromium:
- run all the tests
- retry failures 1
- retry remaining failures 2
- retry remaining failures 3

That way all the flaky tests are retried in the very end of the run. Proposed semantics of `max-failures` would make it useless in case of that strategy.

So I'm going back to understanding your use case better. Could you share more why you need it?

**Comment by dimkin-eu** [comment]
`--retries=1` makes sure, that after 1 retry test will pass
`--max-failures=1` makes sure, that after 1 test failed, the suite fails

as for now, each failure counts, even if the test will pass after retry

so if n tests will fail, but be passing after retries, there must be 0 failures 
if 1 test from that n will not pass after all configured retries- this must be +1 to the failures count
( with some extra option, for sure )

**Comment by pavelfeldman** [comment]
I understand the rationale, but I'd like to understand the use case as well. Let's say we introduce `--max-hard-failures=1`, but will only retry tests in the end of the test run, would that help solving your use case?

**Comment by dimkin-eu** [comment]
kinda fail fast, but slower :)
as specs are running in parallel with limited workers it will anyway save some time

**Comment by pavelfeldman** [comment]
Hang on, with the new strategy, we will run 1000 tests first (no fail fast yet), then we will re-run them all once (no fail fast) and then we re-run them third time (now you'll get fail fast-ish). So is you say `--max-hard-failures` you end up running 10010 tests instead of 10015, which does not sounds like a fail fast to me.

**Comment by dimkin-eu** [comment]
from another perspective - 
10 specs, with 10 tests in each, running using 10 browsers -> 1k tests
with `--retries=1` it passes in 99.9999%, but if it fails in 1st spec, the suite will run till the end - all 1k tests
with `--max-failures=1` suite will be failed immediately, (but if the failed test will be retried the whole suite will pass)

**Comment by pavelfeldman** [comment]
We had a good conversation offline, arriving at setting max-failures to something like 20 - that is a handful of flaky tests that have failed. It would still allow early recovery from the broken infrastructure and would allow us to push retries into the end of the run.

**Comment by pavelfeldman** [comment]
Closing as per above, please feel free to open a new issue if this does not cover your use case.

**Comment by cgetzen** [comment]
>  We had a good conversation offline, arriving at setting max-failures to something like 20 - that is a handful of flaky tests that have failed. It would still allow early recovery from the broken infrastructure and would allow us to push retries into the end of the run.

Hi @pavelfeldman, I don't understand how this helps when there is a "true" test failure and the total tests failure count is < 20: the suite will fail, but run to completion (no "fail-fast").

**Comment by debasisj** [comment]
just a question -`max-failures` will overwrite any `retries`? 
For example in my test suite `--max-failures=1` set as 1 but i want to retry a single test multiple times .. but as soon we have first failure - no retries happens

**Comment by JJ034** [comment]
> just a question -`max-failures` will overwrite any `retries`? For example in my test suite `--max-failures=1` set as 1 but i want to retry a single test multiple times .. but as soon we have first failure - no retries happens

Yes, `--max-failures` overrides retries. In terms of running a single test multiple times you can use `--repeat-each=n`. If used in conjunction with `--retries=0` then it will always run `n` times whether the tests pass or fail.
If doing `--repeat-each=n` with `--retries=2` for example, the test will still run at least `n` times with each individual run of the test having up to 2 retries.
